undertow/src/main/java/org/wildfly/extension/undertow/deployment/UndertowDeploymentProcessor.java

```diff
@@ -303,15 +303,17 @@ private void processDeployment(final WarMetaData warMetaData, final DeploymentUn
         }
 
         final SecurityMetaData securityMetaData = deploymentUnit.getAttachment(ATTACHMENT_KEY);
-        if (isVirtualDomainRequired(deploymentUnit) || isVirtualMechanismFactoryRequired(deploymentUnit)) {
-            securityDomain = builder.requires(securityMetaData.getSecurityDomain());
-        } else if(securityDomainName != null) {
+        if(securityDomainName != null) {
             if (mappedSecurityDomain.test(securityDomainName)) {
                 applySecurityFunction = builder.requires(capabilitySupport.getCapabilityServiceName(Capabilities.CAPABILITY_APPLICATION_SECURITY_DOMAIN, securityDomainName));
             } else {
                 throw ROOT_LOGGER.deploymentConfiguredForLegacySecurity();
             }
         }
+        else if (isVirtualDomainRequired(deploymentUnit) || isVirtualMechanismFactoryRequired(deploymentUnit)) {
+            securityDomain = builder.requires(securityMetaData.getSecurityDomain());
+        }
+
         if (isVirtualMechanismFactoryRequired(deploymentUnit)) {
```

Comment:
I think we can move this if statement into the else if:

```
} else if (isVirtualDomainRequired(deploymentUnit) || isVirtualMechanismFactoryRequired(deploymentUnit)) {
        securityDomain = builder.requires(securityMetaData.getSecurityDomain());

        if (isVirtualMechanismFactoryRequired(deploymentUnit)) {
            ...
        }
}

Reply:
done

Reply:
I don't see this change here.

Reply:
done